File: src/lib/configuration/loader.ts
import { parse } from "@std/yaml";
import { exists } from "@std/fs";
import { dirname, join } from "@std/path";
import { ConfigurationError } from "./base.ts";

/**
 * Configuration file loader with environment support
 */
export class ConfigurationLoader {
  private static readonly CONFIG_DIR = ".jiji";
  private static readonly CONFIG_EXTENSIONS = [".yml", ".yaml"];
  private static readonly DEFAULT_CONFIG_FILES = [
    "deploy",
  ];

  /**
   * Searches for configuration files in the file system
   */
  static async findConfigFile(
    environment?: string,
    startPath: string = Deno.cwd(),
  ): Promise<string | null> {
    const configFilenames = this.buildConfigFilenames(environment);
    let currentPath = startPath;

    while (true) {
      for (const filename of configFilenames) {
        const configPath = join(currentPath, this.CONFIG_DIR, filename);
        if (await exists(configPath)) {
          return configPath;
        }
      }

      const parentPath = dirname(currentPath);
      if (parentPath === currentPath) {
        // Reached filesystem root
        break;
      }
      currentPath = parentPath;
    }

    return null;
  }

  /**
   * Loads configuration from a file
   */
  static async loadFromFile(
    configPath: string,
  ): Promise<Record<string, unknown>> {
    try {
      if (!await exists(configPath)) {
        throw new ConfigurationError(
          `Configuration file not found: ${configPath}`,
        );
      }

      const content = await Deno.readTextFile(configPath);
      const parsed = parse(content);

      if (!parsed || typeof parsed !== "object") {
        throw new ConfigurationError(
          "Configuration file must contain a valid YAML object",
        );
      }

      return parsed as Record<string, unknown>;
    } catch (error) {
      if (error instanceof ConfigurationError) {
        throw error;
      }

      throw new ConfigurationError(
        `Failed to load configuration from ${configPath}: ${
          error instanceof Error ? error.message : String(error)
        }`,
      );
    }
  }

  /**
   * Loads configuration with environment support
   */
  static async loadConfig(
    environment?: string,
    configPath?: string,
    startPath?: string,
  ): Promise<{ config: Record<string, unknown>; path: string }> {
    let actualConfigPath: string;

    if (configPath) {
      // Use provided path
      actualConfigPath = configPath;
    } else {
      // Search for config file
      const foundPath = await this.findConfigFile(environment, startPath);
      if (!foundPath) {
        throw new ConfigurationError(
          this.buildConfigNotFoundMessage(environment),
        );
      }
      actualConfigPath = foundPath;
    }

    const config = await this.loadFromFile(actualConfigPath);
    return { config, path: actualConfigPath };
  }

  /**
   * Builds possible configuration filenames for an environment
   */
  private static buildConfigFilenames(environment?: string): string[] {
    const filenames: string[] = [];

    // Environment-specific files first
    if (environment) {
      for (const base of this.DEFAULT_CONFIG_FILES) {
        for (const ext of this.CONFIG_EXTENSIONS) {
          filenames.push(`${base}.${environment}${ext}`);
        }
      }
      // Also try just the environment name
      for (const ext of this.CONFIG_EXTENSIONS) {
        filenames.push(`${environment}${ext}`);
      }
    }

    // Default files
    for (const base of this.DEFAULT_CONFIG_FILES) {
      for (const ext of this.CONFIG_EXTENSIONS) {
        filenames.push(`${base}${ext}`);
      }
    }

    return filenames;
  }

  /**
   * Builds error message for missing configuration
   */
  private static buildConfigNotFoundMessage(environment?: string): string {
    const envStr = environment ? ` for environment '${environment}'` : "";
    const exampleFiles = environment
      ? [
        `.jiji/deploy.${environment}.yml`,
        `.jiji/deploy.${environment}.yaml`,
        `.jiji/${environment}.yml`,
        `.jiji/${environment}.yaml`,
        ".jiji/deploy.yml",
        ".jiji/deploy.yaml",
      ]
      : [".jiji/deploy.yml", ".jiji/deploy.yaml"];

    return `No jiji configuration file found${envStr}. ` +
      `Please create one of the following files:\n${
        exampleFiles.map((f) => `  - ${f}`).join("\n")
      }`;
  }

  /**
   * Validates that a configuration path is readable
   */
  static async validateConfigPath(path: string): Promise<void> {
    try {
      const stat = await Deno.stat(path);
      if (!stat.isFile) {
        throw new ConfigurationError(
          `Configuration path is not a file: ${path}`,
        );
      }
    } catch (error) {
      if (error instanceof Deno.errors.NotFound) {
        throw new ConfigurationError(
          `Configuration file not found: ${path}`,
        );
      }
      if (error instanceof Deno.errors.PermissionDenied) {
        throw new ConfigurationError(
          `Permission denied reading configuration file: ${path}`,
        );
      }
      throw error;
    }
  }

  /**
   * Gets all available configuration files in a directory
   */
  static async getAvailableConfigs(
    searchPath: string = Deno.cwd(),
  ): Promise<string[]> {
    const configs: string[] = [];
    const configDir = join(searchPath, this.CONFIG_DIR);

    try {
      if (!await exists(configDir)) {
        return configs;
      }

      for await (const entry of Deno.readDir(configDir)) {
        if (entry.isFile && this.isConfigFile(entry.name)) {
          configs.push(join(configDir, entry.name));
        }
      }
    } catch {
      // Ignore errors when reading directory
    }

    return configs.sort();
  }

  /**
   * Checks if a filename is a valid configuration file
   */
  private static isConfigFile(filename: string): boolean {
    return this.CONFIG_EXTENSIONS.some((ext) => filename.endsWith(ext));
  }

  /**
   * Extracts environment name from configuration filename
   */
  static extractEnvironment(configPath: string): string | undefined {
    const basename = configPath.split("/").pop() || "";
    const nameWithoutExt = basename.replace(/\.(yml|yaml)$/, "");

    // Check if it's environment-specific (e.g., deploy.production.yml)
    const parts = nameWithoutExt.split(".");
    if (parts.length === 2 && this.DEFAULT_CONFIG_FILES.includes(parts[0])) {
      return parts[1];
    }

    // Check if it's just an environment name (e.g., production.yml)
    if (
      parts.length === 1 && !this.DEFAULT_CONFIG_FILES.includes(parts[0])
    ) {
      return parts[0];
    }

    return undefined;
  }

  /**
   * Merges multiple configuration objects
   */
  static mergeConfigs(
    base: Record<string, unknown>,
    ...overrides: Record<string, unknown>[]
  ): Record<string, unknown> {
    const result = { ...base };

    for (const override of overrides) {
      for (const [key, value] of Object.entries(override)) {
        if (
          value && typeof value === "object" && !Array.isArray(value) &&
          result[key] && typeof result[key] === "object" &&
          !Array.isArray(result[key])
        ) {
          // Deep merge for objects
          result[key] = this.mergeConfigs(
            result[key] as Record<string, unknown>,
            value as Record<string, unknown>,
          );
        } else {
          // Replace for primitives and arrays
          result[key] = value;
        }
      }
    }

    return result;
  }
}

File: crates/jiji-config/src/loader.rs
use crate::error::ConfigError;
use crate::schema::Config;
use std::fs;
use std::path::{Path, PathBuf};

const CONFIG_DIR: &str = ".jiji";
const CONFIG_EXTENSIONS: [&str; 2] = ["yml", "yaml"];
const DEFAULT_CONFIG_BASE: &str = "deploy";

/// Ordered candidate filenames (checked in this priority order, at every directory level).
fn build_config_filenames(environment: Option<&str>) -> Vec<String> {
    let mut names = Vec::new();
    if let Some(env) = environment {
        for ext in CONFIG_EXTENSIONS {
            names.push(format!("{DEFAULT_CONFIG_BASE}.{env}.{ext}"));
        }
        for ext in CONFIG_EXTENSIONS {
            names.push(format!("{env}.{ext}"));
        }
    }
    for ext in CONFIG_EXTENSIONS {
        names.push(format!("{DEFAULT_CONFIG_BASE}.{ext}"));
    }
    names
}

/// The path `jiji init` writes to for a given environment.
pub fn build_config_path(environment: Option<&str>) -> PathBuf {
    let filename = match environment {
        Some(env) => format!("{DEFAULT_CONFIG_BASE}.{env}.yml"),
        None => format!("{DEFAULT_CONFIG_BASE}.yml"),
    };
    Path::new(CONFIG_DIR).join(filename)
}

/// Search upward from `start`, checking each candidate filename (in priority order) at every
/// directory level before climbing to the parent. Returns the first match.
pub fn find_config_file(environment: Option<&str>, start: &Path) -> Option<PathBuf> {
    let filenames = build_config_filenames(environment);
    let mut current = start.to_path_buf();
    loop {
        for filename in &filenames {
            let candidate = current.join(CONFIG_DIR).join(filename);
            if candidate.is_file() {
                return Some(candidate);
            }
        }
        match current.parent() {
            Some(parent) => current = parent.to_path_buf(),
            None => break,
        }
    }
    None
}

fn build_config_not_found_message(environment: Option<&str>) -> String {
    let env_str = environment
        .map(|e| format!(" for environment '{e}'"))
        .unwrap_or_default();
    let example_files: Vec<String> = build_config_filenames(environment)
        .iter()
        .map(|f| format!("  - {CONFIG_DIR}/{f}"))
        .collect();
    format!(
        "No jiji configuration file found{env_str}. Please create one of the following files:\n{}",
        example_files.join("\n")
    )
}

/// Reads and parses a config file. Does not run semantic validation (see `crate::validation`).
pub fn load_from_file(path: &Path) -> Result<Config, ConfigError> {
    if !path.is_file() {
        return Err(ConfigError::FileNotFound(path.display().to_string()));
    }
    let content = fs::read_to_string(path)?;
    let value: serde_yaml::Value =
        serde_yaml::from_str(&content).map_err(|source| ConfigError::Load {
            path: path.display().to_string(),
            source,
        })?;
    if !value.is_mapping() {
        return Err(ConfigError::NotAnObject);
    }
    serde_yaml::from_value(value).map_err(|source| ConfigError::Load {
        path: path.display().to_string(),
        source,
    })
}

/// Loads config either from an explicit path or by searching upward from `start`.
pub fn load_config(
    environment: Option<&str>,
    config_path: Option<&Path>,
    start: &Path,
) -> Result<(Config, PathBuf), ConfigError> {
    let actual_path = match config_path {
        Some(p) => p.to_path_buf(),
        None => find_config_file(environment, start)
            .ok_or_else(|| ConfigError::NotFound(build_config_not_found_message(environment)))?,
    };
    let config = load_from_file(&actual_path)?;
    Ok((config, actual_path))
}

/// Lists `.jiji/*.{yml,yaml}` files directly under `search_path` (non-recursive: unlike
/// `find_config_file`, this does not climb parent directories). Missing/unreadable dir -> `[]`.
pub fn get_available_configs(search_path: &Path) -> Vec<PathBuf> {
    let config_dir = search_path.join(CONFIG_DIR);
    let mut configs = Vec::new();
    if let Ok(entries) = fs::read_dir(&config_dir) {
        for entry in entries.flatten() {
            let path = entry.path();
            let is_config_file = path
                .file_name()
                .and_then(|n| n.to_str())
                .map(|name| {
                    CONFIG_EXTENSIONS
                        .iter()
                        .any(|ext| name.ends_with(&format!(".{ext}")))
                })
                .unwrap_or(false);
            if path.is_file() && is_config_file {
                configs.push(path);
            }
        }
    }
    configs.sort();
    configs
}
